File: src/main/java/com/danyazero/utils/GenerationContext.java
package com.danyazero.utils;

import com.danyazero.model.MethodSign;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.MethodVisitor;

import java.lang.reflect.Field;
import java.util.*;

public class GenerationContext {
    private final MethodSign currentMethodSign;
    private final ClassWriter cw;
    private final Map<String, String> imports;
    private final Map<MethodSign, MethodContext> methodSigns;

    public GenerationContext(Map<String, String> imports) {
        this.currentMethodSign = null;
        this.cw = new ClassWriter(0);
        this.imports = imports;
        this.methodSigns = new HashMap<>();
    }

    public GenerationContext(MethodSign currentMethodSign, GenerationContext generationContext) {
        this.currentMethodSign = currentMethodSign;
        this.cw = generationContext.getClassWriter();
        this.imports = generationContext.imports;
        this.methodSigns = generationContext.methodSigns;
    }

    public ClassWriter getClassWriter() {
        return cw;
    }

    public GenerationContext newMethodContext(MethodSign methodSign, MethodVisitor methodVisitor) {
        methodSigns.put(methodSign, new MethodContext(methodVisitor));

        return new GenerationContext(methodSign, this);
    }

    public GenerationContext newMethodContext(MethodSign methodSign) {
        methodSigns.put(methodSign, new MethodContext());

        return new GenerationContext(methodSign, this);
    }

    public MethodVisitor getMethodVisitor(MethodSign methodSign) {
        if (methodSigns.containsKey(methodSign)) {
            return methodSigns.get(methodSign).getMethodVisitor();
        }
        throw new RuntimeException("Method Context Not Found: " + methodSign);
    }

    public MethodVisitor getMethodVisitor() {
        if (methodSigns.containsKey(this.currentMethodSign)) {
            return methodSigns.get(this.currentMethodSign).getMethodVisitor();
        }
        throw new RuntimeException("Method Context Not Found: " + this.currentMethodSign);
    }

    public MethodContext getMethodContext() {
        if (methodSigns.containsKey(currentMethodSign)) {
            return methodSigns.get(currentMethodSign);
        }
        throw new RuntimeException("Method Context Not Found: " + currentMethodSign);
    }

    public GenerationContext getGenerationContext(MethodSign methodSign) {
        return new GenerationContext(methodSign, this);
    }

    public void addImport(String importName) {
        var simpleName = importName.split("\\.");
        var lastToken = simpleName.length - 1;

        this.imports.put(importName, simpleName[lastToken]);
    }

    public String resolveImport(String importName) {
        if (imports.containsKey(importName)) return imports.get(importName);

        return this.resolveCoreImport(importName)
                .orElseThrow(() -> new RuntimeException("No such import: " + importName));
    }

    private Optional<String> resolveCoreImport(String importName) {
        try {
            return Optional.of(
                    Class.forName("java.lang." + importName).getName()
            );
        } catch (Exception e) {
            return Optional.empty();
        }
    }

    public Field resolveClassField(String className, String fieldName) {

        try {
            return Class.forName(className).getField(fieldName);
        } catch (Exception e) {
            throw new RuntimeException("An error occurred while trying to resolve " + className + "->" + fieldName, e);
        }
    }

    public MethodSign getCurrentMethodSign() {
        return currentMethodSign;
    }
}

File: src/main/kotlin/utils/GenerationContext.kt
package com.danyazero.utils

import com.danyazero.model.Type
import com.danyazero.model.VariableInfo
import com.danyazero.node.Signature
import org.objectweb.asm.ClassWriter
import org.objectweb.asm.MethodVisitor
import java.lang.reflect.Field

class GenerationContext(
    var className: String? = null,
    private var methodVisitor: MethodVisitor? = null,
    private val classWriter: ClassWriter? = null,
    private val imports: MutableMap<String, String> = mutableMapOf(),
    private var scopeContext: ScopeContext? = null,
    private val classContext: MutableMap<String, Signature> = HashMap()
) {
    fun resolveMethod(owner: String, signature: Signature) : Signature {
        val methodSignature = methodSignature(owner, signature)
        return classContext[methodSignature] ?: throw RuntimeException("method with signature - $methodSignature not found")
    }

    fun defineMethod(signature: Signature) : String {
        val methodSignature = methodSignature(className ?: throw RuntimeException("Class name not defined"), signature)
        println("Definition of method with signature - $methodSignature")
        classContext[methodSignature] = signature

        return methodSignature
    }

    private fun methodSignature(owner: String, signature: Signature) : String = """$owner.${signature.getSignature(this)}"""

    fun enterScope() {
        scopeContext?.enterScope() ?: throw RuntimeException("Failed enter scope: Scope context is null")
    }

    fun exitScope() {
        scopeContext?.exitScope() ?: throw RuntimeException("Failed exit scope: Scope context is null")
    }

    fun defineVariable(name: String, isConstant: Boolean = false, type: Type<*>): Short {
        return scopeContext?.defineVariable(name, isConstant, type)
            ?: throw RuntimeException("Failed define variable: Scope context is null")
    }

    fun resolveVariable(name: String): VariableInfo? {
        return scopeContext?.resolveVariable(name)
            ?: throw RuntimeException("Failed resolve variable: Scope context is null")
    }

    fun resolveLocalVariable(name: String): VariableInfo? {
        return scopeContext?.resolveLocalVariable(name)
            ?: throw RuntimeException("Failed resolve variable: Scope context is null")
    }


    fun addImport(importName: String) {
        val simpleName = importName.split("\\.".toRegex()).toTypedArray().last()

        imports[simpleName] = importName
    }

    fun resolveImport(importName: String): String {
        if (imports.containsKey(importName)) {
            return imports[importName] ?: throw RuntimeException("No such import: $importName")
        }

        return this.resolveCoreImport(importName) ?: throw RuntimeException("No such import: $importName")
    }

    private fun resolveCoreImport(importName: String): String? {
        return try {
            java.lang.Class.forName("java.lang.$importName").name
        } catch (_: Exception) {
            null
        }
    }

    fun resolveClassField(className: String?, fieldName: String): Field {
        try {
            return java.lang.Class.forName(className).getField(fieldName)
        } catch (e: Exception) {
            throw RuntimeException("An error occurred while trying to resolve $className.$fieldName", e)
        }
    }

    fun getMethodVisitor(): MethodVisitor {
        return methodVisitor ?: throw RuntimeException("An error occurred while getting method visitor.")
    }

    fun setMethodVisitor(methodVisitor: MethodVisitor, signature: Signature) {
        this.methodVisitor = methodVisitor
        this.scopeContext = ScopeContext()
        this.classContext[signature.getSignature(this)] = signature
    }

    fun getClassWriter(): ClassWriter {
        return classWriter ?: throw RuntimeException("An error occurred while getting class writer visitor.")
    }
}